pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package backupsink
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/backup/backuppb"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/admission"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+type SSTSinkKeyWriter struct {
+	FileSSTSink
+	// prevKey represents the last key written using WriteKey. When writing a new
+	// key, this helps determine if the last key written was mid-row. This resets
+	// on each flush.
+	prevKey roachpb.Key
+
+	// Caching the targetFileSize from the cluster settings to avoid multiple
+	// lookups during writes.
+	targetFileSize int64
+}
+
+func MakeSSTSinkKeyWriter(
+	conf SSTSinkConf, dest cloud.ExternalStorage, pacer *admission.Pacer,
+) *SSTSinkKeyWriter {
+	return &SSTSinkKeyWriter{
+		FileSSTSink:    *MakeFileSSTSink(conf, dest, pacer),
+		targetFileSize: targetFileSize.Get(conf.Settings),
+	}
+}
+
+// WriteKey writes a single key to the SST file. The key should be the full key,
+// including the prefix. Reset needs to be called prior to WriteKey whenever
+// writing keys from a new span. Keys must also be written in order.
+//
+// Flush should be called after the last key is written to ensure that the SST
+// is written to the destination.
+func (s *SSTSinkKeyWriter) WriteKey(ctx context.Context, key storage.MVCCKey, value []byte) error {
+	if len(s.flushedFiles) == 0 {
+		return errors.AssertionFailedf(
+			"no BackupManifest_File to write key to, call Reset before WriteKey",
+		)
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
```

Comment:
nit: just call `s.flushedFiles[len(s.flushedFiles)-1].Span.ContainsKey()` to avoid the local variable